### PR TITLE
[#4873] feat(core): support list group

### PR DESCRIPTION
**Pull request by LiuQhahah**
### What changes were proposed in this pull request?

(Please outline the changes and how this PR fixes the issue.)

### Why are the changes needed?

support list group

Fix: #4873

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

UT

### CODE REVIEW

**Comment by jerqi** [comment]
You can refer to the pull request https://github.com/apache/gravitino/pull/4055/files

**Comment by jerqi** [comment]
https://github.com/apache/gravitino/pull/4055 @LiuQhahah This pull request about listing users have been merged. Do you have time to rework of this pull request?

**Comment by LiuQhahah** [comment]
> @LiuQhahah #4055 pull request about listing users have been merged. Do you have time to rework this pull request?

Hi @jerqi 

what action need i to do ?
just close the PR?

**Comment by jerqi** [comment]
> > @LiuQhahah #4055 pull request about listing users have been merged. Do you have time to rework this pull request?
> 
> Hi @jerqi
> 
> what action need i to do ? just close the PR?

#4055 is the pull request to support to list users.
You pull request is to list groups. You can refer to #4055 to modify your pull request.

There are several points to modify.
1. You should add more uts and integration test (You can see AccessControlIT.java).
2. You should support listGroupNames and listGroup operations. 
3.

**Comment by LiuQhahah** [comment]
> > > @LiuQhahah #4055 pull request about listing users have been merged. Do you have time to rework this pull request?
> > 
> > 
> > Hi @jerqi
> > what action need i to do ? just close the PR?
> 
> #4055 is the pull request to support to list users. You pull request is to list groups. Two operations are very similar. You can refer to #4055 to modify your pull request.
> 
> There are several points to modify.
> 
> 1. You should add more uts and integration test (You can see AccessControlIT.java).
> 2. You should support listGroupNames and listGroup operations.

Understood, thanks for the advice!
I've been a little busy for a while, so I'll start working on this task today.

**Comment by jerqi** [comment]
@LiuQhahah Do you have time to continue this work? Because we want to release this feature in 0.6.1. The 0.6.1 will be released at the end of this month.

**Comment by LiuQhahah** [comment]
> @LiuQhahah Do you have time to continue this work? Because we want to release this feature in 0.6.1. The 0.6.1 will be released at the end of this month.

Hi @jerqi 
I'll submit a version tonight.

**Comment by jerqi** [comment]
> > @LiuQhahah Do you have time to continue this work? Because we want to release this feature in 0.6.1. The 0.6.1 will be released at the end of this month.
> 
> Hi @jerqi I'll submit a version tonight.

I have updated partial user list implement. You can see https://github.com/apache/gravitino/pull/4894

File: clients/client-java/src/main/java/org/apache/gravitino/client/GravitinoClient.java [hunk 1]
```diff
@@ -227,6 +227,15 @@ public Group getGroup(String group) throws NoSuchGroupException, NoSuchMetalakeE
     return getMetalake().getGroup(group);
   }
 
+
```
[the remark this answers is not in the document]
Comment (LiuQhahah) [reply clients/client-java/src/main/java/org/apache/gravitino/client/GravitinoClient.java:230]: I'll add to that, thanks.

**Comment by jerqi** [comment]
@LiuQhahah Could you give the permission of your repo? We are eager to finish the feature before 0.6.1.

**Comment by LiuQhahah** [comment]
> @LiuQhahah Could you give the permission of your repo? I can help you finish this pull request. We are eager to finish the feature before 0.6.1.

Okay, thanks.


image

File: core/src/main/java/org/apache/gravitino/authorization/UserGroupManager.java [hunk 2]
```diff
@@ -197,4 +198,30 @@ Group getGroup(String metalake, String group) {
...
+      LOG.error("Listing group under metalake {} failed due to storage issues", metalake, ioe);
+      throw new RuntimeException(ioe);
+    }
+  }
+
+  String[] listGroupNames(String metalake) {
+    return Arrays.stream(listGroupInternal(metalake, false))
+        .map(Group::name)
+        .toArray(String[]::new);
+  }
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/authorization/UserGroupManager.java:226]: Can you please move this method above, after `listGroups`.

File: clients/client-java/src/test/java/org/apache/gravitino/client/TestUserGroup.java [hunk 3]
```diff
@@ -327,6 +330,58 @@ public void testRemoveGroups() throws Exception {
...
+        new String[] {"group1", "group2"}, gravitinoClient.listGroupNames());
+    ErrorResponse errRespNoMetaLake =
+        ErrorResponse.notFound(NoSuchMetalakeException.class.getSimpleName(), "metalake not found");
+    buildMockResource(Method.GET, groupPath, null, errRespNoMetaLake, SC_NOT_FOUND);
+    Exception ex =
+        Assertions.assertThrows(
+            NoSuchMetalakeException.class,
+            () -> {
+              gravitinoClient.listGroupNames();
+            });
```
[the remark this answers is not in the document]
Comment (jerqi) [reply clients/client-java/src/test/java/org/apache/gravitino/client/TestUserGroup.java]: OK.

File: core/src/main/java/org/apache/gravitino/authorization/UserGroupManager.java (hunk 2 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/authorization/UserGroupManager.java]: OK.

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/provider/base/GroupMetaBaseSQLProvider.java [hunk 4]
```diff
@@ -34,6 +36,40 @@ public String selectGroupIdBySchemaIdAndName(
...
   }
 
+  public String listGroupPOsByMetalake(@Param("metalakeName") String metalakeName) {
+    return "SELECT gt.group_id as groupId, gt.group_name as groupName, gt.metalake_id as metalakeId,"
+        + " gt.audit_info as auditInfo, gt.current_version as currentVersion, gt.last_version as lastVersion,"
+        + " gt.deleted_at as deletedAt FROM "
+        + GROUP_TABLE_NAME
+        + " gt JOIN "
+        + MetalakeMetaMapper.TABLE_NAME
+        + " mt ON gt.metalake_id = mt.metalake_id  WHERE mt.metalake_name = #{metalakeName}"
```
Comment (yuqi1129) [inline core/src/main/java/org/apache/gravitino/storage/relational/mapper/provider/base/GroupMetaBaseSQLProvider.java:46]: excessive space before WHERE

File: common/src/main/java/org/apache/gravitino/dto/responses/GroupListResponse.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  @Override
+  public void validate() throws IllegalArgumentException {
+    super.validate();
+
+    Preconditions.checkArgument(groups != null, "groups must not be null");
+    Arrays.stream(groups)
+        .forEach(
+            group -> {
+              Preconditions.checkArgument(
+                  StringUtils.isNotBlank(group.name()), "group 'name' must not be null and empty");
```
Comment (yuqi1129) [inline common/src/main/java/org/apache/gravitino/dto/responses/GroupListResponse.java:69]: null or blank

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/provider/base/GroupMetaBaseSQLProvider.java (hunk 4 above)
Comment (jerqi) [reply core/src/main/java/org/apache/gravitino/storage/relational/mapper/provider/base/GroupMetaBaseSQLProvider.java:46]: OK, removed.

File: common/src/main/java/org/apache/gravitino/dto/responses/GroupListResponse.java (hunk 5 above)
Comment (jerqi) [reply common/src/main/java/org/apache/gravitino/dto/responses/GroupListResponse.java]: Fixed.

**Review by jerryshao — APPROVED** [review]

**Comment by jerryshao** [comment]
@jerqi this PR cannot be backported automatically, please help to do it manually.

**Pull request merged**